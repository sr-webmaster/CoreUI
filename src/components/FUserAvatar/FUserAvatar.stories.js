import { MAIN } from '../../../.storybook/categories'
import FUserAvatar from './FUserAvatar.vue'
import { withKnobs, number, select, text } from '@storybook/addon-knobs'

export default {
  title: `${MAIN}|FUserAvatar`,
  id: 'FUserAvatar',
  decorators: [withKnobs]
}

const FIXTURES = {
  COLORS: {
    primary: 'primary',
    secondary: 'secondary',
    accent: 'accent'
  }
}

/**
 * https://storybook.js.org/docs/guides/guide-vue/#step-4-write-your-stories
 */
export const Empty = () => ({
  components: {
    FUserAvatar
  },
  data: () => ({
    user: null
  }),
  template: `
    <f-user-avatar user="user"></f-user-avatar>
  `
})

export const NoUser = () => ({
  components: {
    FUserAvatar
  },
  props: {
    size: {
      default: number('Size', 48)
    },
    color: {
      default: select('Color', FIXTURES.COLORS, 'primary')
    }
  },
  data: () => ({
    user: null
  }),
  template: `
    <f-user-avatar :size="size" :user="user" :color="color"></f-user-avatar>
  `
})

export const Initials = () => ({
  components: {
    FUserAvatar
  },
  props: {
    size: {
      default: number('Size', 48)
    },
    color: {
      default: select('Color', FIXTURES.COLORS, 'primary')
    }
  },
  data: () => ({
    user: {
      first_name: 'Tim',
      last_name: 'Baio'
    }
  }),
  template: `
    <f-user-avatar :size="size" :user="user" :color="color"></f-user-avatar>
  `
})

export const Image = () => ({
  components: {
    FUserAvatar
  },
  props: {
    size: {
      default: number('Size', 48)
    },
    color: {
      default: select('Color', FIXTURES.COLORS, 'primary')
    }
  },
  data: () => ({
    user: {
      first_name: 'Tim',
      last_name: 'Baio',
      avatar: text('Image', 'https://placeimg.com/800/600/people'),
      name: 'Tim Baio'
    }
  }),
  template: `
    <f-user-avatar :size="size" :user="user" :color="color"></f-user-avatar>
  `
})
