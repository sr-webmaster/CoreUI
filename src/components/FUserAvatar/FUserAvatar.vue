<template>
  <v-avatar
    :size="size"
    :color="color"
  >
    <img
      v-if="hasImage"
      :src="image"
      :alt="user.name"
    >
    <span
      v-else-if="hasName"
      :class="color + '-text--text'"
      :style="initialLetterStyle"
    >
      {{ initials }}
    </span>
    <v-icon
      v-else
      :color="color + '-text'"
      :size="size"
    >
      account_circle
    </v-icon>
  </v-avatar>
</template>

<script>
import get from 'lodash/get'
import isEmpty from 'lodash/isEmpty'
export default {
  name: 'FUserAvatar',
  props: {
    size: {
      type: Number,
      default: 48
    },
    color: {
      type: String,
      default: 'accent'
    },
    user: {
      type: Object,
      default: () => ({})
    }
  },
  computed: {
    hasName () {
      return !isEmpty(this.initials)
    },
    initials () {
      const firstName = get(this, 'user.first_name[0]', '')
      const lastName = get(this, 'user.last_name[0]', '')
      return (firstName + lastName).toUpperCase()
    },
    initialLetterStyle () {
      return 'font-size:' + (this.size / 2) + 'px'
    },
    image () {
      return get(this, 'user.avatar', '')
    },
    hasImage () {
      return !isEmpty(this.image)
    }
  }
}
</script>
