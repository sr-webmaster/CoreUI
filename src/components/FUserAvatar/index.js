import Component from './FUserAvatar'
export default Component
