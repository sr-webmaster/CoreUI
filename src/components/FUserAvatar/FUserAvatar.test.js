import { mount, shallowMount } from '@vue/test-utils'
import createLocalVue from 'vue-cli-plugin-freshinup-ui/utils/testing/createLocalVue'
import Component from '@freshinup/core-ui/src/components/FUserAvatar'
import { Initials, NoUser, Image } from './FUserAvatar.stories'

describe('FUserAvatar', () => {
  describe('Computed', () => {
    describe('initials', () => {
      test('returns empty string', () => {
        const wrapper = shallowMount(Component)
        const vm = wrapper.vm
        expect(vm.initials).toEqual('')
        wrapper.setProps({
          user: {}
        })
        expect(vm.initials).toEqual('')
        wrapper.setProps({
          user: {
            first_name: '',
            last_name: ''
          }
        })
        expect(vm.initials).toEqual('')
      })
      test('returns initials of the first and last name', () => {
        const wrapper = shallowMount(Component, {
          propsData: {
            user: {
              first_name: 'Tim',
              last_name: 'Baio'
            }
          }
        })
        const vm = wrapper.vm
        expect(vm.initials).toEqual('TB')
      })
    })
    describe('image', () => {
      test('returns empty string', () => {
        const wrapper = shallowMount(Component)
        const vm = wrapper.vm
        expect(vm.image).toEqual('')
        wrapper.setProps({
          user: {}
        })
        expect(vm.image).toEqual('')
        wrapper.setProps({
          user: {
            avatar: ''
          }
        })
        expect(vm.image).toEqual('')
      })
      test('returns avatar image path', () => {
        const wrapper = shallowMount(Component, {
          propsData: {
            user: {
              avatar: 'me.jpg'
            }
          }
        })
        const vm = wrapper.vm
        expect(vm.image).toEqual('me.jpg')
      })
    })
    describe('hasImage', () => {
      test('returns false', () => {
        const wrapper = shallowMount(Component)
        const vm = wrapper.vm
        expect(vm.hasImage).toEqual(false)
        wrapper.setProps({
          user: {}
        })
        expect(vm.hasImage).toEqual(false)
        wrapper.setProps({
          user: {
            avatar: ''
          }
        })
        expect(vm.hasImage).toEqual(false)
      })
      test('returns true', () => {
        const wrapper = shallowMount(Component, {
          propsData: {
            user: {
              avatar: 'me.jpg'
            }
          }
        })
        const vm = wrapper.vm
        expect(vm.hasImage).toEqual(true)
      })
    })
  })

  describe('Methods', () => {
    // none
  })

  describe('Visuals', () => {
    test('No User', () => {
      // Use mount and snapshot assertions
      const wrapper = mount(NoUser(), {
        localVue: createLocalVue().localVue
      })

      expect(wrapper.isVueInstance()).toBe(true)
      expect(wrapper.element).toMatchSnapshot()
    })

    test('Initials', () => {
      // Use mount and snapshot assertions
      const wrapper = mount(Initials(), {
        localVue: createLocalVue().localVue
      })

      expect(wrapper.isVueInstance()).toBe(true)
      expect(wrapper.element).toMatchSnapshot()
    })

    test('Image', () => {
      // Use mount and snapshot assertions
      const wrapper = mount(Image(), {
        localVue: createLocalVue().localVue
      })

      expect(wrapper.isVueInstance()).toBe(true)
      expect(wrapper.element).toMatchSnapshot()
    })
  })
})
