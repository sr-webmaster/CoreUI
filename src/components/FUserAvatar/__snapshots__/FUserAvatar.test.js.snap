// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FUserAvatar Visuals Image 1`] = `
<div
  class="v-avatar primary"
  style="height: 48px; width: 48px;"
>
  <img
    alt="Tim Baio"
    src="https://placeimg.com/800/600/people"
  />
</div>
`;

exports[`FUserAvatar Visuals Initials 1`] = `
<div
  class="v-avatar primary"
  style="height: 48px; width: 48px;"
>
  <span
    class="primary-text--text"
    style="font-size: 24px;"
  >
    
    TB
  
  </span>
</div>
`;

exports[`FUserAvatar Visuals No User 1`] = `
<div
  class="v-avatar primary"
  style="height: 48px; width: 48px;"
>
  <i
    aria-hidden="true"
    class="v-icon material-icons theme--light primary-text--text"
    style="font-size: 48px;"
  >
    account_circle
  </i>
</div>
`;
